lib/jellyfish_web/socket.ex

```diff
@@ -0,0 +1,138 @@
+defmodule JellyfishWeb.Socket do
+  @moduledoc false
+  @behaviour Phoenix.Socket.Transport
+
+  require Logger
+
+  alias Jellyfish.{Room, RoomService}
+
+  @impl true
+  def child_spec(_opts) do
+    # No additional processes are spawned, returning child_spec for dummy task
+    %{id: __MODULE__, start: {Task, :start_link, [fn -> :ok end]}, restart: :transient}
+  end
+
+  @impl true
+  def connect(state) do
+    Logger.info("New incoming WebSocket connection...")
+
+    with {:ok, peer_id} <- Map.fetch(state.params, "peer_id"),
+         {:ok, room_id} <- Map.fetch(state.params, "room_id"),
+         {:ok, room_pid} <- RoomService.find_room(room_id),
+         {:ok, :disconnected} <- Room.get_peer_connection_status(room_pid, peer_id) do
+      state =
+        state
+        |> Map.put(:room_id, room_id)
+        |> Map.put(:peer_id, peer_id)
+
+      Logger.info(
+        "WebSocket connection from peer #{inspect(peer_id)} accepted, room #{inspect(room_id)}"
+      )
+
+      {:ok, state}
+    else
+      {:ok, :connected} ->
+        Logger.warn(
+          "WebSocket connection for peer #{inspect(state.params["peer_id"])} in room #{inspect(state.params["room_id"])} already exists, rejected"
+        )
+
+        :error
+
+      {:error, :room_not_found} ->
+        Logger.warn(
+          "Room #{inspect(state.params["room_id"])} not found, ignoring incoming WebSocket connection"
+        )
+
+        :error
+
+      {:error, :peer_not_found} ->
+        Logger.warn(
+          "Peer #{inspect(state.params["peer_id"])} not found in room #{inspect(state.params["room_id"])}, ignoring incoming WebSocket connection"
+        )
+
+        :error
+
+      :error ->
+        Logger.warn(
+          "No room_id/peer_id in connection params, ignoring incoming WebSocket connection"
+        )
+
+        :error
+    end
+  end
+
+  @impl true
+  def init(state) do
+    with {:ok, room_pid} <- RoomService.find_room(state.room_id),
+         :ok <- Room.set_peer_connected(room_pid, state.peer_id) do
+    else
+      {:error, :room_not_found} ->
+        Logger.error(
+          "Trying to connect signaling from non existent peer #{inspect(state.peer_id)}), room: #{inspect(state.room_id)}"
+        )
+
+      {:error, :peer_not_found} ->
+        Logger.error(
+          "Trying to connect signaling to non existent room #{inspect(state.room_id)})"
+        )
+    end
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_in({encoded_message, [opcode: :text]}, state) do
+    case Jason.decode(encoded_message) do
+      {:ok, %{"type" => "mediaEvent", "data" => data}} ->
+        case RoomService.find_room(state.room_id) do
+          {:ok, room_pid} ->
+            send(room_pid, {:media_event, state.peer_id, data})
+
+          {:error, :room_not_found} ->
+            Logger.warn(
+              "Trying to send Media Event to room #{inspect(state.room_id)}) that does not exists "
+            )
```

Comment:
I don't think this should be possible. Both `init` and `connect` will return an error if the room doesn't exist, so you can just save the pid somewhere

Reply:
But room can be deleted/crash after socket is initialised, but before some media event is received (I think so at least, or is that some nonsense?).